packages/netlify-cms-core/src/valueObjects/EditorComponent.js

```diff
@@ -1,39 +1,35 @@
-import { Record, fromJS } from 'immutable';
+import { fromJS } from 'immutable';
 import { isFunction } from 'lodash';
 
 const catchesNothing = /.^/;
-/* eslint-disable no-unused-vars */
-const EditorComponent = Record({
-  id: null,
-  label: 'unnamed component',
-  icon: 'exclamation-triangle',
-  fields: [],
-  pattern: catchesNothing,
-  fromBlock(match) {
-    return {};
-  },
-  toBlock(attributes) {
-    return 'Plugin';
-  },
-  toPreview(attributes) {
-    return 'Plugin';
-  },
-});
-/* eslint-enable */
+const bind = fn => isFunction(fn) && fn.bind(null);
 
 export default function createEditorComponent(config) {
-  const configObj = new EditorComponent({
-    id: config.id || config.label.replace(/[^A-Z0-9]+/gi, '_'),
-    label: config.label,
-    icon: config.icon,
-    fields: fromJS(config.fields),
-    pattern: config.pattern,
-    fromBlock: isFunction(config.fromBlock) ? config.fromBlock.bind(null) : null,
-    toBlock: isFunction(config.toBlock) ? config.toBlock.bind(null) : null,
-    toPreview: isFunction(config.toPreview)
-      ? config.toPreview.bind(null)
-      : config.toBlock.bind(null),
-  });
+  const {
+    id = null,
+    label = 'unnamed component',
+    icon = 'exclamation-triangle',
+    type = 'shortcode',
+    widget = 'object',
+    pattern = catchesNothing,
+    fields = [],
+    fromBlock,
+    toBlock,
+    toPreview,
+    ...remainingConfig
+  } = config;
 
-  return configObj;
+  return {
+    id: id || label.replace(/[^A-Z0-9]+/gi, '_'),
+    label,
+    type,
+    icon,
+    widget,
+    pattern,
+    fromBlock: bind(fromBlock) || (() => ({})),
+    toBlock: bind(toBlock) || (() => 'Plugin'),
+    toPreview: bind(toPreview) || (!widget && (bind(toBlock) || (() => 'Plugin'))),
+    fields: fromJS(fields),
+    ...remainingConfig,
+  };
```

Comment:
`createEditorComponent()` now accepts a plain config object returns a plain object. Changes to the API are backwards compatible. Most notably, a `widget` property is now accepted, so widgets can more or less _be_ the editor component.